app/src/main/java/com/ahugenb/hra/calculator/CalculatorView.kt

```diff
@@ -0,0 +1,158 @@
+                maxLines = 1,
+                modifier = Modifier.weight(1f, true),
+                value = abv.value,
+                keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Decimal),
+                onValueChange = {
+                    if (it.isSanitized() && it.smartToDouble().isValidPercent()) {
+                        abv.value = it
+                        viewModel.updateCalculation(
+                            abv.value.smartToDouble(),
+                            volume.value.smartToDouble(),
+                            drinks.value.smartToDouble(),
+                            mlChecked.value
+                        )
+                    }
+                },
+                label = { Text(text = "% ABV") }
+            )
+            OutlinedTextField(
+                maxLines = 1,
+                modifier = Modifier.weight(1f, true),
+                value = drinks.value,
+                keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Decimal),
+                onValueChange = {
+                    if (it.isSanitized() && it.smartToDouble().isValidDrinks()) {
+                        drinks.value = it
+                        viewModel.updateCalculation(
+                            abv.value.smartToDouble(),
+                            volume.value.smartToDouble(),
+                            drinks.value.smartToDouble(),
+                            mlChecked.value
+                        )
+                    }
+                },
+                label = { Text(text = "No. of Drinks") }
+            )
+        }
+        Text(text = stringResource(labelId))
+        Switch(
+            checked = mlChecked.value,
+            onCheckedChange = {
+                mlChecked.value = it
+
+                viewModel.updateCalculation(
+                    abv.value.smartToDouble(),
+                    volume.value.smartToDouble(),
+                    drinks.value.smartToDouble(),
+                    mlChecked.value
+                )
+            }
+        )
+
+        val units = viewModel.units.collectAsState().value
+        val pureEthanol = viewModel.pureEthanol.collectAsState().value
+        Text(
+            textAlign = TextAlign.Right,
+            text = stringResource(R.string.hra_units, units), style = MaterialTheme.typography.h4,
+            modifier = Modifier
+                .padding(horizontal = 8.dp)
+                .fillMaxWidth(),
+        )
+        Text(
+            textAlign = TextAlign.Right,
+            text = stringResource(ethanolId, pureEthanol),
+            style = MaterialTheme.typography.h4,
+            modifier = Modifier
+                .padding(horizontal = 8.dp)
+                .fillMaxWidth()
+        )
+    }
+}
+
+/*
+* sanitized input:
+*  -is less than 10 digits in length
+*  -may not contain more than one decimal
+*  -may only contain digits from validInput
+*/
+private fun String.isSanitized(): Boolean {
```

Comment:
IMO the ViewModel should provide the validation instead of the view itself.

Reply:
This is actually view logic
`
                value = volume.value,
`
`                onValueChange = {
                    if (it.isSanitized() && it.smartToDouble().isValidVolume()) {
                        volume.value = it
                        `
This is how we stop them from typing erroneous things into the box at all - because evaluation is done on input not after a button press